core/consensus/babe/impl/babe_impl.cpp

```diff
@@ -144,40 +144,131 @@ namespace kagome::consensus::babe {
+          std::chrono::microseconds(650'000);  // 0.65s (wavm)
+#else
+      auto block_execution = std::chrono::microseconds(50'000);  // 50ms (wavm)
+#endif
+      auto header_loading = std::chrono::microseconds(5'000);        // 5ms
+      auto state_loading = std::chrono::microseconds(1800'000'000);  // 0.5hr
+      auto warp_proportion = 10'000;  // ~one set id change for each 10k blocks
+
+      auto warp_catchup = lag_slots * header_loading  // time of headers loading
+                            / warp_proportion  // part of requesting headers
+                        + state_loading;       // time of state loading
+      auto fast_catchup = lag_slots * header_loading  // time of headers loading
+                        + 512 * block_execution  // execute non-finalized blocks
+                        + state_loading;         // time of state loading
+      auto full_catchup = lag_slots * block_execution;  // execute all blocks
+
+      auto warp_lag = warp_catchup / babe_config_repo_->slotDuration();
+      auto fast_lag = fast_catchup / babe_config_repo_->slotDuration();
+      auto full_lag = full_catchup / babe_config_repo_->slotDuration();
+
+      warp_sync_duration = warp_catchup + warp_lag * block_execution;
+      fast_sync_duration = fast_catchup + fast_lag * block_execution;
+      full_sync_duration = full_catchup + full_lag * block_execution;
+    }
 
-      const auto &state_root = best_block_header.state_root;
+    bool allow_warp_sync_for_auto = false; // should it select warp for auto
 
-      // Check if target block has state
-      if (auto res = trie_storage_->getEphemeralBatchAt(state_root);
-          res.has_error()) {
+    // Check if target block does not have state (full sync not available)
+    bool full_sync_available = true;
+    if (auto res = trie_storage_->getEphemeralBatchAt(state_root);
+        not res.has_value()) {
+      if (sync_method_ == SyncMethod::Full) {
         SL_WARN(log_, "Can't get state of best block: {}", res.error());
         SL_CRITICAL(log_,
                     "Try restart at least once with `--sync Fast' CLI arg");
         return false;
       }
+      full_sync_available = false;
+    }
+
+    switch (sync_method_) {
+      case SyncMethod::Auto:
+        if (full_sync_duration < fast_sync_duration and full_sync_available) {
+          SL_INFO(log_, "Sync mode auto: decided Full sync");
+          sync_method_ = SyncMethod::Full;
+        } else if (fast_sync_duration < warp_sync_duration
+                   or not allow_warp_sync_for_auto) {
+          SL_INFO(log_, "Sync mode auto: decided Fast sync");
+          sync_method_ = SyncMethod::Fast;
+        } else {
+          SL_INFO(log_, "Sync mode auto: decided Warp sync");
+          sync_method_ = SyncMethod::Warp;
+        }
+        break;
+
+      case SyncMethod::Full:
+        if (fast_sync_duration < full_sync_duration) {
+          SL_INFO(log_, "Fast sync would be faster then Full such was defined");
```

Comment:
IMO "Fast sync would be faster th*a*n Full sync that was selected" sounds better. If agree, consider to update other log messages as well

Reply:
Ok